src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1953,38 +2001,37 @@ int makeSlot(BoundExpression node)
                             }
                         }
                         break;
-                    case BoundKind.DefaultLiteral:
+
                     case BoundKind.DefaultExpression:
+
+                        if (node.Type is not null)
```

Comment:
The only concern I have about doing this for all default expressions is the possibility of introducing slots for just "normal" usage of default like `M(x, y, z, default(S))`. That could increase memory usage. I *think* we would only create the slot if we are actually updating the null state, though. e.g. doing a null test, warning about a dereference, etc.

I did observe that we actually track a flow state for constants in at least some cases, which makes me feel more OK about tracking the state for `default(object)` and so on. [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLTIgWwJaqnGAe1QAcFCTkAfAY0IDtkYACAAQEYAGAfmYA9mAXmb04AGzEBuALAAoPgDoAKoQDKMBNnoBzABQBKScwD0x5gHcoCelu1zFK9Zp0GjpkYQtWbOuUA===).

Reply:
> The only concern I have about doing this for all default expressions is the possibility of introducing slots for just "normal" usage of default like `M(x, y, z, default(S))`. That could increase memory usage. I _think_ we would only create the slot if we are actually updating the null state, though. e.g. doing a null test, warning about a dereference, etc.

I think creating a slot for normal usages will be needed if we want `NotNull` to work?

```csharp
#nullable enable

using System.Diagnostics.CodeAnalysis;

class C<T>
{
    void AssertNotNull([NotNull] T t) { }

    void M1()
    {
        AssertNotNull(default);
        default(T).ToString(); // ok
    }
}
```